zerver/lib/message_cache.py

```diff
@@ -320,28 +321,9 @@ def get_rendering_realm_id(message: Message) -> int:
 
     @staticmethod
     def ids_to_dict(needed_ids: List[int]) -> List[Dict[str, Any]]:
-        # This is a special purpose function optimized for
-        # callers like get_messages_backend().
-        fields = [
-            "id",
-            DB_TOPIC_NAME,
-            "date_sent",
-            "last_edit_time",
-            "edit_history",
-            "content",
-            "rendered_content",
-            "rendered_content_version",
-            "recipient_id",
-            "recipient__type",
-            "recipient__type_id",
-            "sender_id",
-            "sending_client__name",
-            "sender__realm_id",
-        ]
-        # Uses index: zerver_message_pkey
-        messages = Message.objects.filter(id__in=needed_ids).values(*fields)
-        MessageDict.sew_submessages_and_reactions_to_msgs(messages)
-        return [MessageDict.build_dict_from_raw_db_row(row) for row in messages]
+        return MessageDict.messages_to_dicts(
+            Message.objects.filter(id__in=needed_ids), use_sender_realm=True
+        )
```

Comment:
I'm a little worried about the `use_sender_realm` logic here. I have a vague recollection that previously, we only used the sender's realm in rendering code paths for certain lazily written tests, but this looks like it might make that potentially happen in the main message-fetching code path.

(Given cross-realm bots like notification bot, it can be dangerous to use the sender realm for any processing)

Reply:
I think I was attempting to replicate the existing logic, but I agree that it looks funky.  I've stripped these commits off for now.